translator/src/translator.rs

```diff
@@ -0,0 +1,450 @@
+            prev_scope: 0,
+            class_members: HashMap::new(),
+            in_constructor: false,
+        }
+    }
+
+    pub fn translate(&mut self) {
+        for stmt in self.file.tree.body.iter() {
+            self.visit_stmt(stmt);
+        }
+    }
+
+    pub fn write_indent(&mut self) {
+        write!(self.output, "{}", "  ".repeat(self.indent_level));
+    }
+
+    fn check_type(&self, node: &Node, typ: &PythonType) {
+        assert!(
+            self.checker.get_type(node) == *typ,
+            "type error at {}, expected {} got {}",
+            self.file.get_position(node.start, node.end),
+            typ, self.checker.get_type(node)
+        );
+    }
+
+    fn enter_scope(&mut self, pos: u32) {
+        let symbol_table = self.checker.get_symbol_table(None);
+        self.prev_scope = self.current_scope;
+        self.current_scope = symbol_table.get_scope(pos);
+    }
+
+    fn leave_scope(&mut self) {
+        self.current_scope = self.prev_scope;
+    }
+
+    fn python_type_to_cpp(&self, python_type: &PythonType) -> String {
+        let details;
+        match python_type {
+            PythonType::Class(c) => {
+                details = &c.details;
+            },
+            PythonType::Instance(i) => {
+                details = &i.class_type.details;
+            },
+            _ => {
+                return String::from(format!("<unknown type {}>", python_type));
+            }
+        };
+        // If the current symbol table already contains details.name,
+        // we do not need to qualify it, otherwise qualify it with the module name,
+        // unless it is a builtin type in which case we do not qualify it.
+        let symbol_table = self.checker.get_symbol_table(None);
+        match symbol_table.lookup_in_scope(&details.name, self.current_scope) {
+            Some(_) => details.name.to_string(),
+            None => {
+                let symbol_table = self.checker.get_symbol_table(Some(details.declaration_path.symbol_table_id));
+                if symbol_table.file_path.as_path().ends_with("builtins.pyi") {
+                    details.name.to_string()
+                } else {
+                    format!("{}::{}", get_module_name(symbol_table.file_path.as_path()), &details.name)
+                }
+            }
+        }
+    }
+}
+
+impl<'a> TraversalVisitor for CppTranslator<'a> {
+    fn visit_stmt(&mut self, s: &ast::Statement) {
+        self.write_indent();
+        match s {
+            Statement::ExpressionStatement(e) => self.visit_expr(e),
+            Statement::Import(i) => self.visit_import(i),
+            Statement::ImportFrom(i) => self.visit_import_from(i),
+            Statement::AssignStatement(a) => {
+                self.visit_assign(a);
+                writeln!(self.output, ";");
+            },
+            Statement::AnnAssignStatement(a) => self.visit_ann_assign(a),
+            Statement::AugAssignStatement(a) => self.visit_aug_assign(a),
```

Comment:
out of curiosity, whats ann and aug assign and how is it different than the default assign?

Reply:
Good question, this was already there -- I assume Ann stands for annotated (with type annotation?) and aug is for things like +=, see also the grammar here https://docs.python.org/3/reference/grammar.html